### PR TITLE
gh-91117: Ensure integer mod and pow operations use cached small ints

### PR DESCRIPTION
**Pull request by sweeneyde**
https://bugs.python.org/issue46961

### CODE REVIEW

**Comment by sweeneyde** [comment]
I found the call graph of function names a lot to juggle, but the relevant call graphs are like this:
```
long_divmod:
    l_divmod:
        fast_floor_div
        fast_mod
        long_divrem (doesn't check for small remainders):
            divrem1 (returns fresh longs)
            x_divrem (returns fresh longs)

long_mod:
    l_mod:
        fast_mod
        long_rem (doesn't check for small remainders):
            rem1
            x_divrem (returns fresh longs)

long_pow:
    l_mod:
        (same as above)
 
```


As I noted in the issue, `x_divrem` always returns new freshly-allocated ints. Its callers are like this:

```
    - long_divrem
        - uses maybe_small_long on quotient
        - doesn't check if remainder is small <---- oops
    - long_rem
        - throws away quotient
        - doesn't check if remainder is small <---- oops
    - long_true_divide
        - modifies the quotient (!!!)
        - throws away remainder
```

**Comment by mdickinson** [comment]
@sweeneyde Sorry Dennis. This _is_ on my list of things to review, but I'm not likely to get to it for a few days.

File: Objects/longobject.c [hunk 1]
```diff
@@ -2732,6 +2733,7 @@ long_rem(PyLongObject *a, PyLongObject *b, PyLongObject **prem)
     else {
         /* Slow path using divrem. */
         Py_XDECREF(x_divrem(a, b, prem));
+        *prem = maybe_small_long(*prem);
```
Comment (mdickinson) [inline Objects/longobject.c:2736]: It surprised me initially that it was okay to call `maybe_small_long` on a value that might be `NULL`, but I see that other parts of `Objects/longobject.c` also do that (and it makes sense from a convenience perspective).

**Review by mdickinson — APPROVED** [review]
LGTM!

**Comment by sweeneyde** [comment]
Thanks for the review! Looks like we should wait for Blurb-It to sign its CLA ;)

**Pull request merged**